R/layout.R

```diff
@@ -108,3 +105,110 @@ layout_column_wrap <- function(
   )
 }
 
+#' Bootstrap grid layout
+#'
+#' TODO: describe me
+#'
+#' @export
+layout_columns <- function(
+    ...,
+    widths = NA,
+    # TODO: add offsets argument?
+    gap = "1rem",
+    fill = TRUE,
+    fillable = TRUE,
+    height = NULL,
+    height_mobile = NULL,
+    class = NULL
+  ) {
+
+  args <- rlang::list2(...)
+  argnames <- rlang::names2(args)
+
+  attribs <- args[nzchar(argnames)]
+  children <- dropNulls(args[!nzchar(argnames)])
+  n_kids <- length(children)
+
+  # If no widths info is supplied, define number of columns as the number of elements
+  # TODO: maybe it's useful to control n_cols?
+  n_cols <- if (all(is.na(widths))) n_kids else 12
+
+  if (!is_breakpoints(widths)) {
+    widths <- breakpoints(md = widths)
+  }
+
+  width_class <- function(w, breakpoint) {
+    if (length(w) > n_kids) {
+      rlang::warn("Too many `widths` provided; truncating")
+    }
+
+    w <- rep_len(w, n_kids)
+
+    idx_na <- which(is.na(w))
+    if (length(idx_na) > 0) {
+      remaining <- n_cols - sum(w, na.rm = TRUE)
+      w[idx_na] <- max(1, floor(remaining / length(idx_na)))
+    }
+
+    paste0("g-col-", breakpoint, "-", w)
+  }
+
+  width_classes <- do.call(paste,
+    Map(width_class, widths, names(widths))
+  )
+
+  children <- Map(
+    function(el, w, f) {
+      div(
+        class = paste0("g-col-", n_cols),
+        class = w,
+        class = "bslib-gap-spacing",
+        if (f) as_fillable_container(),
+        el
+      )
+    }, children, width_classes, fillable
+  )
+
+  tag <- div(
+    class = "grid bslib-grid",
+    style = css(
+      "--bs-gap" = gap,
+      "--bs-columns" = n_cols,
+      # Always provide the `height:auto` default so that the CSS variable
+      # doesn't get inherited in a scenario like layout_column_wrap(height=200, ..., layout_column_wrap(...))
+      "--bslib-grid-height" = validateCssUnit(height %||% "auto"),
+      "--bslib-grid-height-mobile" = validateCssUnit(height_mobile %||% "auto"),
+      # For some reason, Bootstrap sets `grid-template-rows: 1fr` by default, which is
+      # problematic for a multi-row filling layout. This fixes it...
+      # > page_fillable(layout_columns(c(12, 12), plotly::plot_ly(), plotly::plot_ly()))
+      grid_auto_rows = "1fr"
```

Comment:
So adding `grid-auto-rows` is enough? That's cool.

I can also see some value of letting `height` accept similar values as `widths`. We could default to `1fr` but `height` could support similar syntax as `widths` to be able to do something like `height = c(1, 2)` (first row 1/2 as tall as second) or even using `breakpoints()`. wdyt?

Reply:
Ohh yea, that's a neat idea. I think we'll want to reserve `height` for the usual meaning (it's just CSS height), but maybe we also have a `heights` (or `row_heights`?) that works like you're proposing here?

Reply:
yeah good point about naming. Okay, I'll explore this a bit

Reply:
I just realized that `height` isn't be used for its usual meaning, but that's probably a hold-over from specifics of `layout_column_wrap()`. I've implemented `row_heights` and I think we'll remove all of the `--bslib-grid-height` variables here, since `row_heights` takes us in a different path.

Reply:
I implemented `row_heights`, removed `--bslib-grid-height` css vars that were present, but kept `height` as a variable so that it can be passed to the inline style of the layout container. This also gives us a chance to pass it through `validateCssUnit()`, which increase the chance of an error on mistaken usage. (Same for `width`.)